sd: add SDIO stack to Zephyr

danieldegrasse opened the pull request:
This PR contains the following changes:
- SDIO stack implementation for Zephyr
- New driver APIs for SDHC to support SDIO card interrupts
- Updates for i.MX USDHC driver to support SDIO card interrupts, and DDR50 mode used with SDIO

Although this PR has been verified with internal drivers, no SDIO devices current exist in Zephyr that can be supported with this framework.

danieldegrasse committed:
drivers: sdhc: imx_usdhc: Initialize card clock with CMD0

Use CMD0 to detect card initialization, rather than using card power on.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
sd: add SDIO specification definitions

Add SDIO specification definitions, including SDIO commands.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
sd: add SDIO subsystem

add SDIO subsystem code. SDIO subsystem currently supports
card initialization, SDIO read/write, SDIO extended
read/write, and SDIO card interrupts.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
tests: add test for SDIO subsystem

Add test to verify SDIO subsystem. Note that due to the nature of SDIO
cards, this test only reads from common registers and does not verify
extended reads or writes.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
sd: add sdio header file

Add SDIO header file, defining functions supported by SDIO subsystem

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: sdhc: add support for interrupts

add support for card interrupt sources to SDHC drivers. This can be used
for SDIO card interrupts, or to detect card insertion/removal.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: sdhc: imx_usdhc: add support for card interrupts

Add support for card interrupt sources to USDHC driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: sdhc: imx_usdhc: add support for SDIO RW extended command

Add support for CMD53 (read/write extended) to imx USDHC driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: sdhc: imx_usdhc: enable DDR50 mode

Enable support for DDR50 mode within imx usdhc driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
I'm moving this PR out of draft state, any comments are welcome (although the PR will not merge until we have an open source driver that can be used with this stack)

decsny commented on subsys/sd/sd.c:
```diff
@@ -298,3 +298,22 @@ bool sd_is_card_present(const struct device *sdhc_dev)
+}
+
+/* Disable SD card interrupt */
+int sd_disable_interrupt(const struct device *sdhc_dev)
```
what is the point of having these wrapper functions, if the caller still needs to know the sdhc device

danieldegrasse replied:
There really isn't a good reason for them to exist, the more I consider it. I'll drop these from the PR.

mbrunnen commented:
I would like to propose an API which allows to specify the size in blocks instead of bytes. Some functions are using blocks instead of bytes, e.g. `SDIO_IO_Write_Extended`.

```C
/**
 * @brief Write blocks to SDIO fifo
 *
 * Writes blocks to SDIO register, treating it as a fifo. Writes will
 * all be done to same address.
 * @param func: function to write to
 * @param reg: register address of fifo
 * @param data: data to write to fifo
 * @param blocks: Count of blocks of data to write to card
 * @retval 0 write succeeded
 * @retval -EBUSY: card is busy with another request
 * @retval -ETIMEDOUT: card write timed out
 * @retval -EINVAL: Invalid argument
 * @retval -EIO: I/O error
 */
int sdio_write_blocks_fifo(struct sdio_func *func, uint32_t reg, uint8_t *data,
	uint32_t blocks)
{
	int ret;

	if (func->card->type == CARD_SDMMC) {
		LOG_ERR("MMC does not support SDIO commands");
		return -ENOTSUP;
	}

	if (func->num > 7) {
		LOG_ERR("Invalid function number");
		return -EINVAL;
	}

	if (!(func->card->cccr_flags & SDIO_SUPPORT_MULTIBLOCK))
	{
		LOG_ERR("Card does not support multi block mode");
		return -EINVAL;
	}

	ret = k_mutex_lock(&func->card->lock, K_NO_WAIT);

	if (ret) {
		LOG_WRN("Could not get SD card mutex");
		return -EBUSY;
	}

	ret = sdio_io_rw_extended(func->card, SDIO_IO_WRITE, func->num, reg, false, data, blocks,
				  func->block_size);

	k_mutex_unlock(&func->card->lock);
	return ret;
}
```

mbrunnen commented on include/zephyr/sd/sd.h:
```diff
@@ -64,6 +77,8 @@ struct sd_card {
 	enum card_type type; /*!< Card type */
 	uint32_t flags; /*!< Card flags */
+	uint32_t cccr_flags; /*!< SDIO CCCR data */
+	struct sdio_func func0; /*!< Function 0 common card data */
```
```suggestion
    struct sdio_func funcs[CONFIG_SDIO_MAX_IO_FUNCS];
```
Instead of having to maintain the function objects in the application, I would prefer, that the `sd_card` structure has an array with all functions, e.g. with a configurable static size. The function would not have a `sdio_func*` as argument, but a value from the `sdio_func_num` enumeration.

danieldegrasse replied:
My reasoning for delegating the function objects to the application was that the SDIO driver will know how many card functions it needs to access for the SDIO device it is interfacing with. If we use something like `CONFIG_SDIO_MAX_IO_FUNCS`, then we would have a few issues:
- if multiple SDIO drivers are used, who decides the maximum number of IO functions supported?
- if one SDIO driver needs 7 functions and the other needs only 1, we have some wasted space in the SDIO structure.

Is there an issue you are running into when designing around this API that you feel would be solved by storing the SDIO functions in the card structure?

mbrunnen replied:
> Is there an issue you are running into when designing around this API that you feel
> would be solved by storing the SDIO functions in the card structure?
Now I need a wrapper, which manages the SDIO function blocks. Both ways are acceptable from my point of view.

danieldegrasse replied:
Understood- if it is not impacting your ability to design around the API, I'd prefer we continue to delegate allocation of SDIO function blocks to the application. For what it's worth, this is similar to how Linux defines the SDIO API: https://github.com/torvalds/linux/blob/master/include/linux/mmc/sdio_func.h#L144

npal-cy reviewed (commented):
Hi @danieldegrasse, thank you for your effort here. When do you plan to merge this? I am working on host agnostic version of CY43xxx Wi-Fi driver, which also will required the SDIO stack functionality, which you introduced here.

I would like to take your code base, integrate it with driver and do more test.

Regards,
Nazar

danieldegrasse commented:
> I would like to propose an API which allows to specify the size in blocks instead of bytes. Some functions are using blocks instead of bytes, e.g. SDIO_IO_Write_Extended.

Hi @mbrunnen, thanks for the feedback. From your perspective, what is the issue with writing to the card using `sdio_write_fifo`? This function will use block based I/O for the entire transfer if the data length given is a multiple of the SDIO function's block size. Additionally, a function like `sdio_write_blocks_fifo` would have to return `-ENOTSUP` if the card did not support multiblock I/O transfers, whereas `sdio_write_fifo` can handle this transparently.

danieldegrasse commented on include/zephyr/drivers/sdhc.h:
```diff
@@ -242,6 +251,9 @@ __subsystem struct sdhc_driver_api {
+	int (*enable_interrupt)(const struct device *dev,
+				sdhc_interrupt_cb_t callback, void *user_data);
+	int (*disable_interrupt)(const struct device *dev);
 };
```
Hi- I am a bit confused what you are suggesting here. SD host controllers have one interrupt line. However, individual functions on an SDIO device can have interrupts, and each function interrupt is OR'ed together to issue an interrupt to the SD host controller.

Are you saying we should support enabling and disabling interrupts per each function within the SDIO stack? One way we could support this is to have functions like the following in the SDIO stack:
```
int sdio_enable_interrupt(struct sdio_func *func, sdio_interrupt_cb_t callback, void *user_data);
int sdio_disable_interrupt(struct sdio_func *func);
```

These functions could then only call the provided `callback` if an interrupt occurred for the given SDIO function

npal-cy replied:
Sorry for confusing, i mean  that current enable_interrupt/dissable_interrupt function operates only with Card interrupt
There are another interrupts:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/46566584/47aea88f-ad11-476d-be2e-c54f1f1a39bb)

If we do not want to handle them, maybe we can just rename:
enable_interrupt    -> enable_card_interrupt    
dissable_interrupt  -> dissable_card_interrupt

danieldegrasse replied:
Ah I understand- you mean additional interrupts in the host controller itself. Card removal wasn't something I considered since Zephyr does not support hotplugging devices. I'll do the following:
- rework the signature of the SDHC API to enable interrupt callbacks when an SDIO interrupt occurs, as well as when card insertion occurs
- rework the SDIO interrupt API to allow a consumer to register for interrupts from only one function source

danieldegrasse replied:
> rework the SDIO interrupt API to allow a consumer to register for interrupts from only one function source

After some additional consideration, I don't think we need an API for this feature yet. It won't really add anything, since SDHC drivers themselves can only register one callback at at time. Therefore, the SDIO subsystem would need a way to store callbacks for each function, which defeats the point of keeping SDIO function declarations out of the card structure (since we would now need callback pointers for each function). I will update the PR with a new signature for the SDHC API to allow for other card interrupts (like insertion and removal) to issue callbacks

danieldegrasse commented on include/zephyr/drivers/sdhc.h:
```diff
@@ -231,6 +231,15 @@ struct sdhc_host_props {
+ * Function prototype for SDHC card interrupt callback.
+ * @param user_data: User data, set via @ref sdhc_enable_interrupt
+ */
+typedef void (*sdhc_interrupt_cb_t)(const void *user_data);
```
Sure- this is a good point. I'll update this callback signature.

mbrunnen commented:
> > I would like to propose an API which allows to specify the size in blocks instead of bytes. Some functions are using blocks instead of bytes, e.g. SDIO_IO_Write_Extended.
> 
> Hi @mbrunnen, thanks for the feedback. From your perspective, what is the issue with writing to the card using `sdio_write_fifo`? This function will use block based I/O for the entire transfer if the data length given is a multiple of the SDIO function's block size. Additionally, a function like `sdio_write_blocks_fifo` would have to return `-ENOTSUP` if the card did not support multiblock I/O transfers, whereas `sdio_write_fifo` can handle this transparently.

It is not a real issue, I just didn't like the idea to convert blocks to bytes, which are re-converted to blocks.

danieldegrasse committed:
tests: drivers: sdhc: add test for card insertion interrupt

Add test for card insertion interrupt to SDHC test.

The test will now wait for the card insertion interrupt whenever it is
run without a card in the slot. By inserting a card, the user can verify
that card insertion interrupts work correctly, as this should notify the
test and allow it to proceed.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

npal-cy commented:
> Hi @npal-cy, the blocker on integrating this stack into Zephyr is that no upstreamable drivers exist that utilize it. If you can upstream the CY43xxx driver, then this stack could likely be merged. I just want to make sure there is a demonstrated use case of this stack available when it merges.

Hi @danieldegrasse,
CY43xxx wifi driver, is working on top of [ Infineon Wi-Fi Host Driver (WHD)](https://github.com/Infineon/wifi-host-driver#wi-fi-host-driver-whd). CY43xxx wifi driver implements Zephyr HAL layer for WHD, with following functionality:
- Enable/disable SDIO Card interrupt (will use new driver APIs for SDHC to support SDIO card interrupts for this PR)   
- Implementation of SDIO CMD53/52 (will use SDHC sdhc_request function. The core functions sdio_io_rw_direct/sdio_io_rw_extended from SD/SDIO subsystem are static, so i have to SDHC sdhc_request function for WHD implementation).
- Expect to use sd_init() for basic card initialization.

I have a plan to push a draft next week.

npal-cy commented:
Hi @danieldegrasse, i pushed AIROC Wi-Fi driver: https://github.com/zephyrproject-rtos/zephyr/pull/63721

fabiobaltieri commented on subsys/sd/sdio.c:
```diff
@@ -56,23 +63,997 @@ int sdio_send_ocr(struct sd_card *card, uint32_t ocr)
+static int sdio_set_bus_speed(struct sd_card *card)
+{
+	int ret, retries = CONFIG_SD_RETRY_COUNT;
+	int timing = 0;
+	uint8_t target_speed = 0;
```
I think these two don't need to be initialized.

danieldegrasse replied:
Yeah, you're right- removed both

ifyall commented:
@danieldegrasse, we now have a working SDIO implementation using this PRs functionality available in https://github.com/zephyrproject-rtos/zephyr/pull/63721. Are you able to push this PR to completion so we can merge the dependent PR as well?

Thank you in advance,

Ian

danieldegrasse committed:
include: drivers: sdhc: indentation cleanups

Indentation of multiline function declarations wasn't aligned properly.
Clean this up.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented on drivers/sdhc/imx_usdhc.c:
```diff
@@ -833,7 +1012,14 @@ static int imx_usdhc_init(const struct device *dev)
 			return ret;
 		}
+		gpio_init_callback(&data->cd_callback, card_detect_gpio_cb,
+				BIT(cfg->detect_gpio.pin));
```
good old vim :) maybe too simple, I believe I've fixed all these indentation cases though

fabiobaltieri replied:
I say vim is just simple enough 😉, have you set `cinoptions` correctly? I use `set cinoptions=(0,W2s,:0`, absolutely can't recall what any of that means but works mostly right.

danieldegrasse replied:
I didn't even know about cinoptions- was still using `smartindent`. After some quality time with the manpages it seems that `cinoptions=(0` is the trick. Thank you for the heads up!

fabiobaltieri replied:
yeah no prob, I used to use vim defaults for a while, then something something broke the default at some point and had to force myself figuring that one out...

danieldegrasse commented on include/zephyr/drivers/sdhc.h:
```diff
@@ -231,6 +231,30 @@ struct sdhc_host_props {
+ * @param user_data: User data, set via @ref sdhc_enable_interrupt
+ */
+typedef void (*sdhc_interrupt_cb_t)(const struct device *dev, int reason,
+				const void *user_data);
```
indentation is off in this file- I fixed it for the interrupt API functions, and added another commit to this PR fixing indentation for the remainder of the API definitions (as well as removing the casts)

danieldegrasse commented on include/zephyr/drivers/sdhc.h:
```diff
@@ -430,6 +458,64 @@ static inline int z_impl_sdhc_get_host_props(const struct device *dev,
+	sdhc_interrupt_cb_t callback, int sources, void *user_data)
+{
+	const struct sdhc_driver_api *api =
+		(const struct sdhc_driver_api *)dev->api;
```
Good point- fixed

danieldegrasse commented on include/zephyr/sd/sd_spec.h:
```diff
@@ -756,6 +759,156 @@ enum sd_spec_version {
+
+#define SDIO_CCCR_SPEED	0x13 /*!< SDIO bus speed select */
+#define SDIO_CCCR_SPEED_SHS BIT(0) /*!< high speed support */
+#define SDIO_CCCR_SPEED_MASK (0xE) /*!< bus speed select mask*/
```
I don't believe we do. I've dropped the parenthesis on integer constants in this header

danieldegrasse commented:
Removing DNM label as a WiFi driver now exists to validate this stack with- @fabiobaltieri thank you for the review, I believe I've addressed all comments

danieldegrasse commented:
I believe the casts to `const struct sdhc_driver_api` may be required- other subsystems do this, and without these casts CI fails for some tests (I saw tests/lib/cpp/cxx/ on qemu_cortex_a53 specifically) with the following error:
```
/home/danieldegrasse/zephyr/zephyrproject/zephyr/include/zephyr/drivers/sdhc.h:503:50: error: invalid conversion from 'const void*' to 'const sdhc_driver_api*' [-fpermissive]
  503 |         const struct sdhc_driver_api *api = dev->api;
      |                                             ~~~~~^~~
      |                                                  |
      |                                                  const void*
```

fabiobaltieri commented:
> I believe the casts to `const struct sdhc_driver_api` may be required- other subsystems do this, and without these casts CI fails for some tests (I saw tests/lib/cpp/cxx/ on qemu_cortex_a53 specifically) with the following error:
...

That is so weird, `g++` oddness? Wonder why is it doing that and not in other places.

gmarull commented:
> I believe the casts to `const struct sdhc_driver_api` may be required- other subsystems do this, and without these casts CI fails for some tests (I saw tests/lib/cpp/cxx/ on qemu_cortex_a53 specifically) with the following error:
> 
> ```
> /home/danieldegrasse/zephyr/zephyrproject/zephyr/include/zephyr/drivers/sdhc.h:503:50: error: invalid conversion from 'const void*' to 'const sdhc_driver_api*' [-fpermissive]
>   503 |         const struct sdhc_driver_api *api = dev->api;
>       |                                             ~~~~~^~~
>       |                                                  |
>       |                                                  const void*
> ```

When building for C++ you cannot cast a void to something else without an explicit cast, that's why in Zephyr we have things like `const struct sdhc_driver_api *api = (const struct sdhc_driver_api *)dev->api;` on all API headers.

gmarull commented:
FYI https://eli.thegreenplace.net/2009/11/16/void-and-casts-in-c-and-c

fabiobaltieri commented on subsys/sd/sdio.c:
```diff
@@ -56,23 +63,995 @@ int sdio_send_ocr(struct sd_card *card, uint32_t ocr)
+		LOG_WRN("Could not get SD card mutex");
+		return -EBUSY;
+	}
+	ret = sdio_io_rw_extended_helper(func, SDIO_IO_WRITE, reg, true,
+		data, len);
```
I think you should do a pass with the right vim indentation settings to align to the previous open parenthesis and fix all of these, they just look weird, to me at least :-) apart from that good stuff!

danieldegrasse replied:
I think I've caught the remainder of them in latest push- let me know if you see any others

npal-cy commented on subsys/sd/sdio.c:
```diff
@@ -56,23 +63,995 @@ int sdio_send_ocr(struct sd_card *card, uint32_t ocr)
+			return ret;
+		}
+	}
+	cis_ptr = *((uint32_t *)data);
```
here `cis_ptr` will take 4 bytes from `data`, above read only 3 bytes from SDIO_FBR_CIS, so 4-th byte can be garbage (i am observing this issue with airoc driver), so need to clean `data[0..3]`, or better:
```
	uint32_t cis_ptr = 0, num = 0;
	uint8_t tpl_code, tpl_link;
	bool match_tpl = false;

	memset(&func->cis, 0, sizeof(struct sdio_cis));
	/* First find the CIS pointer for this function */
	for (int i = 0; i < 3; i++) {
		ret = sdio_io_rw_direct(func->card, SDIO_IO_READ, SDIO_FUNC_NUM_0,
			SDIO_FBR_BASE(func->num) + SDIO_FBR_CIS + i, 0, data);
		if (ret) {
			return ret;
		}
		cis_ptr |= *data << (i * 8);
	}
```

npal-cy commented on subsys/sd/sdio.c:
```diff
@@ -56,23 +63,995 @@ int sdio_send_ocr(struct sd_card *card, uint32_t ocr)
+	/* Read CIS tuples until we have read all requested CIS tuple codes */
+	do {
+		/* Read tuple code */
+		ret = sdio_io_rw_direct(func->card, SDIO_IO_READ, func->num,
```
CIS info we must read by F0 (SDIO_FUNC_NUM_0):
`ret = sdio_io_rw_direct(func->card, SDIO_IO_READ, func->num,...`  -> `ret = sdio_io_rw_direct(func->card, SDIO_IO_READ, SDIO_FUNC_NUM_0,...`

danieldegrasse replied:
Thanks for pointing this out- the driver I have been testing with locally didn't run into this issue, but Linux does read the CIS tuples for functions from function 0, so we should take the same approach.

npal-cy commented on subsys/sd/sdio.c:
```diff
@@ -56,23 +63,995 @@ int sdio_send_ocr(struct sd_card *card, uint32_t ocr)
+			continue;
+		}
+		/* Read tuple link */
+		ret = sdio_io_rw_direct(func->card, SDIO_IO_READ, func->num,
```
CIS info we must read by F0 (SDIO_FUNC_NUM_0):
`ret = sdio_io_rw_direct(func->card, SDIO_IO_READ, func->num,...`  -> `ret = sdio_io_rw_direct(func->card, SDIO_IO_READ, SDIO_FUNC_NUM_0,...`

npal-cy commented on subsys/sd/sdio.c:
```diff
@@ -56,23 +63,995 @@ int sdio_send_ocr(struct sd_card *card, uint32_t ocr)
+			memset(data, 0, 255);
+			for (int i = 0; i < tpl_link; i++) {
+				ret = sdio_io_rw_direct(func->card, SDIO_IO_READ,
+					func->num, cis_ptr++, 0, data + i);
```
CIS info we must read by F0 (SDIO_FUNC_NUM_0):
`.. func->num,...`  -> `..SDIO_FUNC_NUM_0,...`

npal-cy commented on subsys/sd/sdio.c:
```diff
@@ -56,23 +63,995 @@ int sdio_send_ocr(struct sd_card *card, uint32_t ocr)
+ * @retval -ETIMEDOUT: card I/O timed out
+ * @retval -EIO: I/O error
+ */
+int sdio_enable_func(struct sd_card *card, struct sdio_func *func,
```
Can we split this one in to two?:
- Initialize SDIO function: `int sdio_init_func(struct sd_card *card, struct sdio_func *func, enum sdio_func_num num)`
- Enable SDIO function: 'int sdio_enable_func(struct sdio_func *func, enum sdio_func_num num)'

It will helpful for case as airoc-wifi driver, where enabling sdio functions is handling by low level wifi driver (WHD).
see my local changes of sdio.c in integration branch: https://github.com/npal-cy/zephyr-1/commit/50ad03a0cb9b1ec878ad6133014024cd99058ec8

danieldegrasse replied:
Can you clarify a bit here how splitting these functions would help with porting the AIROC driver? From what I can tell, the CIS issue you encountered will be resolved by the bugfix you suggested above (reading the CIS information from function 0).

npal-cy replied:
as i told before our WHD driver cares about enabling function F1, F2 in whd_bus_sdio_init according to device specific (e.g enabling F2 after FW downloaded to wifi device).  so if i try to enable F2 before WHD_Init, i will got timeout.

The spiting sdio_init_func/sdio_enable_func give me possibility to use CMD52/53 functions without enabling SDIO functions. Enabling SDIO functions will handling in WHD in proper sequences . WHD is separate asset, we do not want to modify it.

Ses my integration with using sdio_init_func: https://github.com/npal-cy/zephyr-1/commit/cba1a13b66b03b5126113aa67b316a7f51654a67

danieldegrasse replied:
Ok, thank you- I understand now. `whd_init` will leverage CMD52 and CMD53 to load the firmware, but SDIO functions cannot be enabled until you load the firmware. I'll split these functions into two sections, so that you can leverage the SDIO stack here fully.

fabiobaltieri reviewed (approved):
Good stuff!

dleach02 commented:
@carlescufi what approvals do we need to get this merged? I'm not sure what the process is here in that the author is also the maintainer in this area so do we wait on the author to also approve since he is the assignee?

fabiobaltieri commented:
I think this could just get merged, it's an extension of an existing subsystem, the author is the maintainer, do we have any requirement for involving the architecture wg? (not that I'm aware of)

kartben commented:
Some late comments from my side but it would be really great to have the `sd/*.h` headers properly "promoted" in the documentation, i.e. define doxygen groups, add them to the relevant parent groups, and add RST pages in the documentation that would include these groups.

danieldegrasse commented:
> @danieldegrasse
Some late comments from my side but it would be really great to have the sd/*.h headers properly "promoted" in the documentation, i.e. define doxygen groups, add them to the relevant parent groups, and add RST pages in the documentation that would include these groups.

This is a good point- would you be willing to create an issue/enhancement request to track this? You can assign me and I'll make the updates needed.
